spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/abstractnn/IModule.scala

```diff
@@ -0,0 +1,523 @@
+
+import scala.reflect.ClassTag
+
+trait NameUtilities[A <: Activity, B <: Activity, T] {
+  self: IModule[A, B, T] =>
+  private var namePostfix = Integer.toHexString(java.util.UUID.randomUUID().hashCode())
+
+  def getNamePostfix : String = namePostfix
+
+  def setNamePostfix(namePostfix : String) : Unit = this.namePostfix = namePostfix
+
+
+  /**
+   * The name of the module
+   */
+  protected var name : String = null
+
+  def hasName: Boolean = name != null
+
+  /**
+   * Set the module name
+   *
+   * @param name
+   * @return
+   */
+  def setName(name : String) : this.type = {
+    this.name = name
+    this
+  }
+
+  /**
+   * Get the module name, default name is className@namePostfix
+   *
+   * @return
+   */
+  def getName() : String = {
+    if (this.name == null) {
+      s"${this.getClass.getSimpleName}${namePostfix}"
+    } else {
+      this.name
+    }
+  }
+
+  def getPrintName(): String = {
+    val postfix = if (name == null) {
+      namePostfix
+    } else {
+      name
+    }
+    s"${this.getClass.getSimpleName}[${postfix}]"
+
+  }
+}
+
+trait GraphUtilities[A <: Activity, B <: Activity, T] {
+   self: IModule[A, B, T] =>
+  /**
+   * Build graph: some other modules point to current module
+   * @param nodes upstream module nodes
+   * @return node containing current module
+   */
+  def inputs(nodes : ModuleNode[T]*): ModuleNode[T] = {
+    val curNode = new ModuleNode[T](this)
+    nodes.foreach(node => {
+      node.add(curNode, Edge())
+    })
+    curNode
+  }
+
+  /**
+   * Build graph: some other modules point to current module
+   * @param nodes upstream module nodes in an array
+   * @return node containing current module
+   */
+  def inputs(nodes : Array[ModuleNode[T]]): ModuleNode[T] = {
+    val curNode = new ModuleNode[T](this)
+    nodes.foreach(node => {
+      node.add(curNode, Edge())
+    })
+    curNode
+  }
+
+  /**
+   * Build graph: some other modules point to current module
+   * @param first distinguish from another inputs when input parameter list is empty
+   * @param nodesWithIndex upstream module nodes and the output tensor index. The start index is 1.
+   * @return node containing current module
+   */
+  def inputs(first: (ModuleNode[T], Int), nodesWithIndex : (ModuleNode[T], Int)*): ModuleNode[T] = {
+    val curNode = new ModuleNode[T](this)
+    first._1.add(curNode, Edge(first._2))
+    nodesWithIndex.foreach(nodeWithIndex => {
+      nodeWithIndex._1.add(curNode, Edge(nodeWithIndex._2))
+    })
+    curNode
+  }
+}
+
+/**
+ * Module is the basic component of a neural network. It forward activities and backward gradients.
+ * Modules can connect to others to construct a complex neural network.
+ *
+ * @tparam A Input data type
+ * @tparam B Output data type
+ * @tparam T Numeric type of parameter(e.g. weight, bias). Only support float/double now
+ */
+abstract class IModule[A <: Activity: ClassTag, B <: Activity: ClassTag, T: ClassTag]
```

Comment:
Why import a new base class here? why not modify on the existing AbstractModule? This touch too much code. And it's huge API change, user customized code will be broken.

Reply:
A simple method is to add inferShape, compile methods to the AbstractModule.

inferShape method by default throw not imp exceptions, if a layer support infer shape, just override this method.

compile method by default is a empty method and different container override this method with different strategy.